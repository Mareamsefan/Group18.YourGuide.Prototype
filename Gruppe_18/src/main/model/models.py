from sqlalchemy import Table, Column, String, Integer, ForeignKey, DATETIME
from sqlalchemy.orm import relationship
from flask_login import UserMixin
import uuid

from Gruppe_18.src.main.database.sql_alchemy import db

tour_account_association = Table(
    'tour_account_association', db.metadata,
    Column('tour_id', String, ForeignKey('Tour.id')),
    Column('account_id', String, ForeignKey('User.id'))
)


class Account(db.Model, UserMixin):
    __tablename__ = "User"
    id = db.Column(db.String, primary_key=True)
    usertype = Column("usertype", String)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)
    tours = relationship("Tour", secondary=tour_account_association, back_populates="participants")

    def get_id(self):
        return str(self.id)

    def __init__(self, usertype, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.usertype = usertype
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"


class Tour(db.Model):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)
    participants = relationship("Account", secondary=tour_account_association, back_populates="tours")

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

    def get_id(self):
        return str(self.id)
