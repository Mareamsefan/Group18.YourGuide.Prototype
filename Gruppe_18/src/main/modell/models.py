from sqlalchemy import Column, String, Integer, DATETIME
import uuid
from Gruppe_18.src.main.database.sql_alchemy import Base


class Account(Base):
    __tablename__ = "User"
    account_id = Column("id", String, primary_key=True)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.account_id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.account_id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"

    def get_id(self):
        return self.id


class Tour(Base):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    tour_id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.tour_id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.tour_id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

