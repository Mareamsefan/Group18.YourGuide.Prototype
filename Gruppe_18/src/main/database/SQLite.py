import sqlite3

connection = sqlite3.connect("YourGuide.db")
connection.row_factory = sqlite3.Row
cursor = connection.cursor()


cursor.execute("INSERT INTO tour (title, destination, duration, cost, pictureURL, "
               "language, MAX_travelers, Booked, Tour_ID) "
               "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
               ("Tour Title", "Destination", 7.5, 500.0, "example.jpg", "English", 20, 5, "12345"))

cursor.execute("INSERT INTO tour (title, destination, duration, cost, pictureURL, "
               "language, MAX_travelers, Booked, Tour_ID) "
               "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
               ("Tour Title", "Destination", 7.5, 500.0, "example.jpg", "English", 20, 5, "12346"))

connection.commit()

cursor.execute("SELECT * FROM tour")
rows = cursor.fetchall()

# Iterate through the rows and print the data
print(rows)
for row in rows:
    print(row['tour_id'], row['destination'])

# Close the cursor and the connection
cursor.close()
connection.close()




