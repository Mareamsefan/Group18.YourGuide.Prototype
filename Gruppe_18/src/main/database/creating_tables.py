import sqlite3

connection = sqlite3.connect("YourGuide.db")
connection.row_factory = sqlite3.Row
cursor = connection.cursor()

cursor.execute("CREATE TABLE tour (title TEXT, destination TEXT, duration REAL, cost REAL, "
               "pictureURL TEXT, language TEXT, MAX_travelers INTEGER, Booked INTEGER, Tour_ID TEXT)")

cursor.execute("CREATE TABLE Account (username TEXT, password TEXT, phonen TEXT, emailAdress TEXT, account_id Integer)")





