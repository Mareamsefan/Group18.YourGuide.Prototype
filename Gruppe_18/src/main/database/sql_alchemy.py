import os

from flask import Flask
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

app = Flask(__name__, template_folder='../templates')
module_path = os.path.dirname(os.path.abspath(__file__))
database_name = os.path.join(module_path, "../YourGuide.db")

app.config['SQLALCHEMY_DATABASE_URI'] = f'sqlite:///{database_name}'
app.secret_key = 'gruppe18'

engine = create_engine(f"sqlite:///{database_name}", echo=True)
Session = sessionmaker(bind=engine)

def get_session():
    engine = create_engine("sqlite:///YourGuide.db", echo=True)

    Session = sessionmaker(bind=engine)

    return Session()

