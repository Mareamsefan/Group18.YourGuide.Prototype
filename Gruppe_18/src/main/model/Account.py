import json
import uuid


class Account:

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress



    def get_id(self):
        return self.id
