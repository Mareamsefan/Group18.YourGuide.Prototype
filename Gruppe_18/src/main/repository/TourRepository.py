from Gruppe_18.src.main.repository.JSONRepository import JSONRepository

list_with_tours =

class TourRepository (JSONRepository):

    def filter_tour_by_location(tour_list, destination):
        filtered_tours = []
        for tour in tour_list:
            if tour['destination'] == destination:
                filtered_tours.append(tour)
        return filtered_tours

    def filter_tour_by_price(tour_list, max_price, min_price):
        filtered_tours = []
        for tour in tour_list:
            if min_price <= tour['cost'] <= max_price:
                filtered_tours.append(tour)
        return filtered_tours