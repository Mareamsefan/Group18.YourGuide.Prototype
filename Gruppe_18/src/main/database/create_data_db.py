import os

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from Gruppe_18.src.main.database.app_config import db

module_path = os.path.dirname(os.path.abspath(__file__))
database_name = os.path.join(module_path, "../YourGuide.db")
engine = create_engine(f"sqlite:///{database_name}", echo=True)

Session = sessionmaker(bind=engine)
db.metadata.create_all(bind=engine)

def get_session(path):
    engine = create_engine(f"sqlite:///{path}", echo=True)

    Session = sessionmaker(bind=engine)

    return Session()
