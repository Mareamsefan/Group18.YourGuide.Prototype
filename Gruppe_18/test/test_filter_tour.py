import os
import uuid

from approvaltests.scrubbers import scrub_all_guids
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from Gruppe_18.src.main.model.models import db

import pytest
from approvaltests import verify, Options

from Gruppe_18.src.main.repository.TourRepository import TourRepository
from Gruppe_18.test.database.database_handler import get_session
import datetime
from Gruppe_18.src.main.model.models import Tour

@pytest.fixture
def tour_re():
    return TourRepository(get_session())


@pytest.fixture()
def tour():
    return Tour(str(uuid.uuid4()), "Welcome to Dubai",
        datetime.date(2020, 10, 15),
        "Dubai",
        4,
        255,
        15,"English",
        "https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg")


@pytest.fixture
def tour_2():
    return Tour(
        str(uuid.uuid4()),
    "Bergen Fjord Exploration",
    datetime.date(2024, 7, 10),
    "Bergen, Norway",
    4,
    4199,
    15,
    "English",
    "https://www.example.com/bergen-fjords.jpg"
)


@pytest.fixture
def tour_3():
    return Tour(
        str(uuid.uuid4()),
    "Discover Oslo's Charm",
    datetime.date(2024, 8, 5),
    "Oslo, Norway",
    2,
    2500,
    10,
    "Norwegian",
    "https://www.example.com/oslo-city.jpg"
)

@pytest.fixture()
def another_tour():
    return Tour("Exploring Parisian Charm",
                datetime.date(2021, 5, 20),
                "Paris",
                3,
                200,
                1,
                "French",
                "https://www.example.com/paris.jpg")

@pytest.fixture()
def sqlalchemy_session(tour_re, tour, tour_2, tour_3):
    module_path = os.path.dirname(os.path.abspath(__file__))
    database_name = os.path.join(module_path, "Test.db")
    engine = create_engine(f"sqlite:///{database_name}", echo=True)

    session = sessionmaker(bind=engine)()

    db.metadata.create_all(bind=engine)

    tour_re.create_tour(tour)
    tour_re.create_tour(tour_2)
    tour_re.create_tour(tour_3)
    yield session

    session.close()
    db.metadata.drop_all(engine)


approval_options = Options().with_scrubber(scrub_all_guids)


def test_if_tour_is_created_saved_and_retrived(tour_re, sqlalchemy_session):
    saved_data = tour_re.get_all_tours()
    verify(saved_data, options=approval_options)


def test_book_tour_with_available_space(sqlalchemy_session, tour_re, another_tour):
    tour = another_tour
    tour_re.create_tour(tour)
    assert tour_re.book_tour(tour) == True


def test_book_tour_when_fully_booked(sqlalchemy_session, tour, tour_re, another_tour):
    tour_re.create_tour(another_tour)
    tour_re.book_tour(another_tour)
    assert tour_re.book_tour(another_tour) == False
'''
def test_book_multiple_tours_with_available_space(tour, tour_rep):
    tour.max_travelers = 2
    assert tour_rep.book_tour(tour) == True
    assert tour_rep.book_tour(tour) == True

def test_book_more_tours_than_available_space(tour, tour_rep):
    tour.max_travelers = 2
    tour.booked = 2
    assert tour_rep.book_tour(tour) == False
'''
def test_tour_description_generation_is_as_expected(tour, tour_rep):
    given_description = tour_rep.get_tour_description(tour)
    expected_description = "This tour will take you to Italy for 4 hours, and is " \
               "offered in English"
    assert given_description == expected_description


def test_tour_description_generation_is_not_as_expected(tour, tour_rep):
    given_description = tour_rep.get_tour_description(tour)
    tour.destination = "Montenegro"
    expected_description = "This tour will take you to Montenegro for 4 hours, and is " \
               "offered in English"
    assert given_description != expected_description


def test_if_filtering_based_on_nothing_returns_all_tours(sqlalchemy_session, tour_re):
    filter_tour = tour_re.filter_combinations('', '', '', '')

    verify(filter_tour, options=approval_options)


def test_if_filtering_based_on_only_destination_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("Dubai", "", "", "")
    verify(filter_tour, options=approval_options)


def test_if_filtering_based_on_only_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "500", "3000", "")
    verify(filter_tour, options=approval_options)

def test_if_filtering_only_on_max_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "", "3000", "")
    verify(filter_tour, options=approval_options)

def test_if_filtering_only_on_min_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "500", "", "")
    verify(filter_tour, options=approval_options)

def test_if_filtering_based_on_only_language_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "", "", "English")
    verify(filter_tour, options=approval_options)


def test_if_filtering_based_on_destination_price_and_language_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("Dubai", "0", "600", "English")
    verify(filter_tour, options=approval_options)
