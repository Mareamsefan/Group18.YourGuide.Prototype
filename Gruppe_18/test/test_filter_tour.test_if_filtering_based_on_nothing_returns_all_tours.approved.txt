[(<guid_0>) Welcome to Dubai Dubai 4 255 English 15 https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg 0, (<guid_1>) Bergen Fjord Exploration Bergen, Norway 4 4199 English 15 https://www.example.com/bergen-fjords.jpg 0, (<guid_2>) Discover Oslo's Charm Oslo, Norway 2 2500 Norwegian 10 https://www.example.com/oslo-city.jpg 0]
